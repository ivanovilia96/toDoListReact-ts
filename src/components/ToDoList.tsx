import React, {ReactDOM, ReactHTML, ReactInstance, ReactNode, useEffect, useState} from 'react';
import {Route, Link, Switch} from "react-router-dom";
import {Button, Dropdown, Input, Menu, message} from 'antd';
import axios from 'axios';
import Modal from "./Modal";


const {TextArea} = Input;

const {Item} = Menu;

interface Task {
    completed: boolean;
    description: string;
    id: string;
    taskName: string;
}

interface Response {
    data: Array<Task>;
}

const completedTaskType = 'completed';
const inProcessTaskType = 'in process';

const buttonType = {
    delete: 'delete',
    view: 'view',
    complete: 'complete'
}

const baseUrl = 'https://5fd21a6eb485ea0016eef6eb.mockapi.io/v1/tasksList/'
type tasksType = typeof completedTaskType | typeof inProcessTaskType


export default function ToDoList(): JSX.Element {
    const [listOfToDo, setListOfToDo] = useState<Array<Task>>([]);
    const [listOfCompletedToDo, setListOfCompletedToDo] = useState<Array<Task>>([]);
    const [whichTasksWeGonnaToSe, setWhichTasksWeGonnaToSe] = useState<tasksType>(inProcessTaskType)
    const [task, setTask] = useState<Task>({
        completed: false,
        description: '',
        id: '',
        taskName: '',
    })
    const [taskDescription, setTaskDescription] = useState<{ description: string, taskName: string }>({
        description: "",
        taskName: ""
    })

    useEffect(() => {
        axios.get(baseUrl)
            .then((result: Response) => {
                setListOfToDo(result.data.filter(i => !i.completed))
                setListOfCompletedToDo(result.data.filter(i => i.completed))
                message.success('data fetched')
            }, () => {
                message.error('error on fetch data from the server, call your system administrator', 1000)
            })
    }, [])

    function handleChangeTaskStatus(toDo: { id: string; completed: boolean }): void {
        axios.put(`${baseUrl}${toDo.id}`, {
            ...toDo,
            completed: !toDo.completed
        }).then(res => {
                message.success('status changed')
                if (res.data.completed) {
                    setListOfCompletedToDo([...listOfCompletedToDo, res.data])
                    setListOfToDo(listOfToDo.filter(i => i.id !== res.data.id))
                } else {
                    setListOfToDo([...listOfToDo, res.data])
                    setListOfCompletedToDo(listOfCompletedToDo.filter(i => i.id !== res.data.id))
                }
            }, () => message.error('error on change task status, call your system administrator', 10)
        )
    }

    function handleChangeWhichTasksWeGonnaToSe(): void {
        setWhichTasksWeGonnaToSe(whichTasksWeGonnaToSe === completedTaskType ? inProcessTaskType : completedTaskType)
    }

    function deleteTask(id: string, taskType: tasksType): void {
        axios.delete(`${baseUrl}${id}`)
            .then(res => {
                    message.success('task deleted')
                    if (taskType === inProcessTaskType) {
                        setListOfToDo(listOfToDo.filter(i => i.id !== res.data.id))
                    } else {
                        setListOfCompletedToDo(listOfCompletedToDo.filter(i => i.id !== res.data.id))
                    }
                }, () =>
                    message.error('error on delete task, call your system administrator', 10)
            )
    }

    function handleChangeTaskName(e: { target: { value: string } }): void {
        setTask({...task, taskName: e.target.value})
    }

    function handleChangeTaskDescription(e: { target: { value: string } }): void {
        setTask({...task, description: e.target.value})
    }

    function handleAddNewTask(): void {
        axios.post(baseUrl, task)
            .then((res: { data: Task }) => {
                    message.success('task added')
                    setListOfToDo([...listOfToDo, res.data])
                    setTask({
                            completed: false,
                            description: '',
                            id: '',
                            taskName: '',
                        }
                    )
                }, () => message.error('error on add new task, call your system administrator', 10)
            )
    }

    function handleViewDescriptionChange(task: Task): void {
        const {taskName, description} = task;
        setTaskDescription({
            taskName, description
        })
    }

    const menu: JSX.Element = (
        <Menu>
            <Item onClick={handleChangeWhichTasksWeGonnaToSe}>
                {whichTasksWeGonnaToSe === completedTaskType ? inProcessTaskType : completedTaskType}
            </Item>
        </Menu>
    );

    return (
        <div className='main'>
            <span className='header'>to do list</span>
            <Input size='large'
                   addonAfter={
                       <Button
                           disabled={!task.taskName}
                           onClick={handleAddNewTask}
                       >
                           add
                       </Button>
                   }
                   addonBefore=' add to do '
                   placeholder='what do you gonna do ?' className={'input'}
                   value={task.taskName}
                   onChange={handleChangeTaskName}
            />
            <Dropdown
                className={'change-view'}
                overlay={menu}
                arrow={true}
                trigger={["click"]}
            >
                <Button>{whichTasksWeGonnaToSe === completedTaskType ? completedTaskType : inProcessTaskType} </Button>
            </Dropdown>

            <div className={"to-do"}>
                {task.taskName && <TextArea
                  value={task.description}
                  onChange={handleChangeTaskDescription}
                  placeholder={'add here your description for new task'}
                />}
                {whichTasksWeGonnaToSe === inProcessTaskType ? listOfToDo.map(toDo => {
                    const {id, taskName} = toDo;
                    return (
                        <div
                            key={id}
                            className={'task'}
                        >
                            <span className={'task-name'}>  {taskName}</span>
                            <Button
                                className={buttonType.complete}
                                onClick={handleChangeTaskStatus.bind({}, toDo,)}
                            >
                                task complete
                            </Button>

                            <Link to={'/description/tasks/' + taskName} className={buttonType.view}>
                                <Button

                                    onClick={handleViewDescriptionChange.bind({}, toDo)}
                                >
                                    view description
                                </Button>
                            </Link>

                            <Button
                                onClick={deleteTask.bind({}, id, inProcessTaskType)}
                                className={buttonType.delete}
                            >
                                delete
                            </Button>
                        </div>)
                }) : listOfCompletedToDo.map(toDo => {
                    const {id, taskName} = toDo;
                    return (
                        <div key={id}
                             className={'task'}>
                            <span className={'task-name'}>  {taskName}</span>

                            <Button
                                onClick={handleChangeTaskStatus.bind({}, toDo,)}
                                className={buttonType.complete}
                            >
                                move in incomplete
                            </Button>
                            <Link to={'/description/tasks/' + taskName} className={buttonType.view}>
                                <Button
                                    onClick={handleViewDescriptionChange.bind({}, toDo)}
                                >
                                    view description
                                </Button>
                            </Link>
                            <Button
                                onClick={deleteTask.bind({}, id, completedTaskType)}
                                className={buttonType.delete}
                            >
                                delete
                            </Button>
                        </div>
                    )
                })}
            </div>
            <Switch>
                <Route path={`/description/tasks/`}>
                    <Modal
                        title={taskDescription.taskName}
                        description={taskDescription.description}
                    />
                </Route>
            </Switch>
        </div>
    );
}
