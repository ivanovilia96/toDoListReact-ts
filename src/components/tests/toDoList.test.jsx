import React from "react";
import ToDoList, {baseUrl} from "../ToDoList";
import {render} from "enzyme";
import {BrowserRouter as Router} from "react-router-dom";
import axios from "axios";
import {message} from "antd";


describe('ToDoList', () => {
  it("check render: snapshot test", () => {
    const wrapper = render(
      <Router>
        <ToDoList/>
      </Router>);
    expect(wrapper).toMatchSnapshot()
  });

  it.skip("check render: snapshot test after useEffect", () => {
    const mockUseStateChangeFunc = jest.fn()
    jest.spyOn(React, 'useEffect').mockImplementationOnce((arg) => {
      arg()
    })
    const useStateChecker = jest.spyOn(React, 'useState')
      .mockImplementation(initialState => ([initialState, mockUseStateChangeFunc]))

    const successListener = jest.spyOn(message, 'success').mockImplementation(() => {
    })
    const get = jest.spyOn(axios, 'get').mockImplementation(() => Promise.resolve({
      data: [
        {
          completed: true,
          description: 'mock-descr-1',
          id: '1',
          taskName: 'mock-task-name1',
        }
      ]
    }))
    const wrapper = render(
      <Router>
        <ToDoList/>
      </Router>);


    expect(wrapper).toMatchSnapshot()
    expect(get).toHaveBeenCalledWith(baseUrl)
    expect(successListener).toHaveBeenCalledWith()
  });

})

