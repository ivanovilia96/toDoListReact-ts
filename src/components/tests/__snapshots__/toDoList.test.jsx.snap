// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList check render: snapshot test 1`] = `
<div
  class="main"
>
  <span
    class="header"
  >
    to do list
  </span>
  <span
    class="ant-input-group-wrapper ant-input-group-wrapper-lg input"
  >
    <span
      class="ant-input-wrapper ant-input-group"
    >
      <span
        class="ant-input-group-addon"
      >
         add to do 
      </span>
      <input
        class="ant-input ant-input-lg"
        placeholder="what do you gonna do ?"
        type="text"
        value=""
      />
      <span
        class="ant-input-group-addon"
      >
        <button
          class="ant-btn"
          disabled=""
          type="button"
        >
          <span>
            add
          </span>
        </button>
      </span>
    </span>
  </span>
  <button
    class="ant-btn ant-dropdown-trigger change-view"
    type="button"
  >
    <span>
      in process 
    </span>
  </button>
  <div
    class="to-do"
  />
</div>
`;

exports[`ToDoList check render: snapshot test after useEffect 1`] = `
<div
  class="main"
>
  <span
    class="header"
  >
    to do list
  </span>
  <span
    class="ant-input-group-wrapper ant-input-group-wrapper-lg input"
  >
    <span
      class="ant-input-wrapper ant-input-group"
    >
      <span
        class="ant-input-group-addon"
      >
         add to do 
      </span>
      <input
        class="ant-input ant-input-lg"
        placeholder="what do you gonna do ?"
        type="text"
        value=""
      />
      <span
        class="ant-input-group-addon"
      >
        <button
          class="ant-btn"
          disabled=""
          type="button"
        >
          <span>
            add
          </span>
        </button>
      </span>
    </span>
  </span>
  <button
    class="ant-btn ant-dropdown-trigger change-view"
    type="button"
  >
    <span>
      in process 
    </span>
  </button>
  <div
    class="to-do"
  />
</div>
`;
